File: Validations/test_invoice_creation.py
#!/usr/bin/env python3
"""
End-to-End Invoice Creation Test Script (Python Version)
Purpose: Create a complete invoice from customer to database

Test Flow:
1. Find or create customer
2. Find or create product
3. Create invoice with items
4. Verify invoice in database
"""

import requests
import json
import sys
from datetime import datetime
from decimal import Decimal

# Configuration
API_BASE = "https://pharma-backend-production-0c09.up.railway.app/api"
ORG_ID = "11111111-1111-1111-1111-111111111111"

# Color codes for terminal output
class Colors:
    RESET = '\033[0m'
    GREEN = '\033[92m'
    RED = '\033[91m'
    YELLOW = '\033[93m'
    BLUE = '\033[94m'
    BOLD = '\033[1m'

def log(message, level='info'):
    """Print colored log messages"""
    prefixes = {
        'success': f"{Colors.GREEN}✅",
        'error': f"{Colors.RED}❌",
        'warning': f"{Colors.YELLOW}⚠️",
        'info': f"{Colors.BLUE}ℹ️",
        'step': f"{Colors.BOLD}📍"
    }
    print(f"{prefixes.get(level, '')} {message}{Colors.RESET}")

def load_invoice_data(filename='sample_invoice_input.json'):
    """Load invoice data from JSON file"""
    try:
        with open(filename, 'r') as f:
            data = json.load(f)
        log(f"Loaded invoice data from {filename}", 'success')
        return data
    except FileNotFoundError:
        log(f"File {filename} not found", 'error')
        return None
    except json.JSONDecodeError as e:
        log(f"Invalid JSON in {filename}: {e}", 'error')
        return None

def find_or_create_customer(invoice_data):
    """Find existing customer or create new one"""
    log("Step 1: Customer Setup", 'step')
    
    customer_id = invoice_data.get('customer_id')
    if customer_id:
        log(f"Using provided customer ID: {customer_id}", 'info')
        return customer_id
    
    # Search for customer by phone
    phone = invoice_data.get('primary_phone', '')
    if phone:
        try:
            response = requests.get(
                f"{API_BASE}/customers",
                params={"search": phone, "limit": 1},
                headers={"X-Org-Id": ORG_ID},
                timeout=10
            )
            
            if response.status_code == 200:
                data = response.json()
                customers = data.get('customers', [])
                if customers:
                    customer_id = customers[0].get('customer_id')
                    log(f"Found existing customer: ID {customer_id}", 'success')
                    return customer_id
        except Exception as e:
            log(f"Customer search failed: {e}", 'warning')
    
    # Create new customer
    log("Creating new customer...", 'info')
    customer_data = {
        "customer_name": invoice_data.get('customer_name', 'Test Customer'),
        "customer_type": invoice_data.get('customer_type', 'retail'),
        "primary_phone": invoice_data.get('primary_phone', '9999999999'),
        "primary_email": invoice_data.get('primary_email', ''),
        "state": invoice_data.get('state', 'Maharashtra'),
        "city": invoice_data.get('city', 'Mumbai'),
        "credit_limit": invoice_data.get('credit_limit', 50000),
        "credit_period_days": invoice_data.get('credit_period_days', 30)
    }
    
    try:
        response = requests.post(
            f"{API_BASE}/customers",
            json=customer_data,
            headers={
                "X-Org-Id": ORG_ID,
                "Content-Type": "application/json"
            },
            timeout=10
        )
        
        if response.status_code in [200, 201]:
            result = response.json()
            customer_id = result.get('customer_id') or result.get('id') or 1
            log(f"Created new customer: {customer_data['customer_name']} (ID: {customer_id})", 'success')
            return customer_id
        else:
            log(f"Customer creation failed: {response.status_code}", 'warning')
            log(f"Response: {response.text[:200]}", 'warning')
            return 1  # Default fallback
            
    except Exception as e:
        log(f"Customer creation error: {e}", 'error')
        return 1

def create_invoice(invoice_data):
    """Create invoice via API"""
    log("Step 2: Invoice Creation", 'step')
    
    # Ensure customer_id is set
    if not invoice_data.get('customer_id'):
        invoice_data['customer_id'] = find_or_create_customer(invoice_data)
    
    # Add required fields if missing
    invoice_data.setdefault('invoice_date', datetime.now().isoformat())
    invoice_data.setdefault('invoice_type', 'tax_invoice')
    invoice_data.setdefault('payment_method', 'cash')
    invoice_data.setdefault('payment_terms', 'cash')
    invoice_data.setdefault('place_of_supply', 'Maharashtra')
    
    # Calculate totals if not provided
    if 'items' in invoice_data and invoice_data['items']:
        subtotal = 0
        total_tax = 0
        
        for item in invoice_data['items']:
            quantity = float(item.get('quantity', 0))
            unit_price = float(item.get('unit_price', 0))
            discount_percent = float(item.get('discount_percentage', 0))
            gst_percent = float(item.get('gst_percentage', 18))
            
            line_subtotal = quantity * unit_price
            discount_amount = line_subtotal * discount_percent / 100
            taxable = line_subtotal - discount_amount
            tax = taxable * gst_percent / 100
            
            subtotal += line_subtotal
            total_tax += tax
            
            # Add calculated values to item
            item['discount_amount'] = discount_amount
            item['line_total'] = taxable
            item['line_total_with_tax'] = taxable + tax
            
            # Add GST breakup (assuming intra-state)
            if gst_percent > 0:
                item['cgst_amount'] = tax / 2
                item['sgst_amount'] = tax / 2
                item['igst_amount'] = 0
        
        # Set invoice totals
        invoice_data.setdefault('subtotal_amount', subtotal)
        invoice_data.setdefault('total_tax_amount', total_tax)
        invoice_data.setdefault('taxable_amount', subtotal - invoice_data.get('discount_amount', 0))
        
        total = invoice_data['taxable_amount'] + total_tax + invoice_data.get('other_charges', 0)
        invoice_data.setdefault('total_amount', total)
        invoice_data.setdefault('net_amount', total)
        invoice_data.setdefault('final_amount', total)
    
    # Display invoice summary
    print(f"\n{Colors.BOLD}Invoice Summary:{Colors.RESET}")
    print(f"  Customer: {invoice_data.get('customer_name', 'N/A')}")
    print(f"  Items: {len(invoice_data.get('items', []))} product(s)")
    print(f"  Subtotal: ₹{invoice_data.get('subtotal_amount', 0):.2f}")
    print(f"  Tax: ₹{invoice_data.get('total_tax_amount', 0):.2f}")
    print(f"  Total: ₹{invoice_data.get('total_amount', 0):.2f}")
    
    # Send invoice to API
    log("Sending invoice to backend...", 'info')
    
    try:
        response = requests.post(
            f"{API_BASE}/invoices",
            json=invoice_data,
            headers={
                "X-Org-Id": ORG_ID,
                "Content-Type": "application/json"
            },
            timeout=30
        )
        
        log(f"Response status: {response.status_code}", 'info')
        
        if response.status_code in [200, 201]:
            try:
                result = response.json()
                log("Invoice created successfully!", 'success')
                print(f"\n{Colors.BOLD}Invoice Details:{Colors.RESET}")
                print(f"  Invoice Number: {result.get('invoice_number', 'N/A')}")
                print(f"  Invoice ID: {result.get('invoice_id', 'N/A')}")
                print(f"  Status: {result.get('invoice_status', 'Created')}")
                return result
            except json.JSONDecodeError:
                log("Invoice likely created but response parsing failed", 'warning')
                print(f"Raw response: {response.text[:200]}")
                return {'status': 'partial_success'}
        else:
            log(f"Invoice creation failed: HTTP {response.status_code}", 'error')
            error_text = response.text[:500]
            print(f"Error: {error_text}")
            
            # Check for specific errors
            if 'column' in error_text and 'does not exist' in error_text:
                log("Database schema issue detected", 'error')
                print("\nPossible fixes:")
                print("1. Run database migration scripts")
                print("2. Check column names match schema")
                print("3. Verify database connection")
            
            return None
            
    except requests.exceptions.Timeout:
        log("Request timed out", 'error')
        return None
    except requests.exceptions.ConnectionError:
        log("Could not connect to backend", 'error')
        return None
    except Exception as e:
        log(f"Unexpected error: {e}", 'error')
        return None

def verify_invoice(invoice_id):
    """Verify invoice was saved to database"""
    log("Step 3: Invoice Verification", 'step')
    
    if not invoice_id:
        log("No invoice ID to verify", 'warning')
        return False
    
    try:
        response = requests.get(
            f"{API_BASE}/invoices/{invoice_id}",
            headers={"X-Org-Id": ORG_ID},
            timeout=10
        )
        
        if response.status_code == 200:
            log("Invoice verified in database", 'success')
            return True
        else:
            log(f"Could not verify invoice: HTTP {response.status_code}", 'warning')
            return False
            
    except Exception as e:
        log(f"Verification failed: {e}", 'warning')
        return False

def main():
    """Main execution"""
    print(f"\n{Colors.BOLD}🧪 INVOICE CREATION TEST (PYTHON){Colors.RESET}")
    print("=" * 60)
    
    # Get input file from command line or use default
    if len(sys.argv) > 1:
        input_file = sys.argv[1]
    else:
        input_file = 'sample_invoice_input.json'
    
    # Load invoice data
    invoice_data = load_invoice_data(input_file)
    if not invoice_data:
        log("Failed to load invoice data", 'error')
        sys.exit(1)
    
    # Create invoice
    print("\n" + "-" * 60)
    result = create_invoice(invoice_data)
    
    # Verify if created
    if result and result.get('invoice_id'):
        print("\n" + "-" * 60)
        verify_invoice(result['invoice_id'])
    
    # Final status
    print("\n" + "=" * 60)
    if result:
        log("TEST COMPLETED SUCCESSFULLY", 'success')
        print("\n✨ Invoice has been created!")
    else:
        log("TEST FAILED - Check errors above", 'error')
        print("\n⚠️  Invoice creation failed")
        print("\nTroubleshooting:")
        print("1. Check if backend is running: https://pharma-backend-production-0c09.up.railway.app/health")
        print("2. Verify database has required columns (gst_percentage, line_total, etc.)")
        print("3. Check if invoice_items table exists in sales schema")
        print("4. Ensure customer can be created/found")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: Validations/test_invoice_creation.js
#!/usr/bin/env node
/**
 * End-to-End Invoice Creation Test Script
 * Purpose: Validate complete invoice creation workflow from customer to invoice
 * 
 * This script tests:
 * 1. Customer creation/lookup
 * 2. Product verification
 * 3. Invoice creation with items
 * 4. Tax calculations
 * 5. Database persistence
 */

const fs = require('fs');
const path = require('path');

// Configuration
const API_BASE = 'https://pharma-backend-production-0c09.up.railway.app/api';
const ORG_ID = '11111111-1111-1111-1111-111111111111';

// Color codes for console output
const colors = {
  reset: '\x1b[0m',
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  bold: '\x1b[1m'
};

function log(message, type = 'info') {
  const prefix = {
    success: `${colors.green}✅`,
    error: `${colors.red}❌`,
    warning: `${colors.yellow}⚠️`,
    info: `${colors.blue}ℹ️`,
    step: `${colors.bold}📍`
  };
  
  console.log(`${prefix[type] || ''}  ${message}${colors.reset}`);
}

async function createInvoiceFromJSON(inputFile) {
  log('Starting End-to-End Invoice Creation Test', 'step');
  console.log('=' .repeat(60));
  
  // Load input data
  let invoiceData;
  try {
    const inputPath = inputFile || path.join(__dirname, 'sample_invoice_input.json');
    const rawData = fs.readFileSync(inputPath, 'utf8');
    invoiceData = JSON.parse(rawData);
    log(`Loaded invoice data from: ${inputPath}`, 'success');
  } catch (error) {
    log(`Failed to load input file: ${error.message}`, 'error');
    return null;
  }
  
  // Display invoice summary
  console.log('\n' + colors.bold + 'Invoice Summary:' + colors.reset);
  console.log(`Customer: ${invoiceData.customer_name} (${invoiceData.primary_phone})`);
  console.log(`Items: ${invoiceData.items.length} product(s)`);
  console.log(`Payment: ${invoiceData.payment_method || 'cash'}`);
  
  if (invoiceData.items && invoiceData.items.length > 0) {
    console.log('\nProducts:');
    invoiceData.items.forEach((item, index) => {
      console.log(`  ${index + 1}. ${item.product_name} - Qty: ${item.quantity} @ ₹${item.unit_price}`);
    });
  }
  
  // Calculate totals
  const subtotal = invoiceData.subtotal_amount || 0;
  const discount = invoiceData.discount_amount || 0;
  const tax = (invoiceData.cgst_amount || 0) + (invoiceData.sgst_amount || 0) + (invoiceData.igst_amount || 0);
  const other = invoiceData.other_charges || 0;
  const total = invoiceData.net_amount || invoiceData.total_amount || 0;
  
  console.log('\nAmount Breakdown:');
  console.log(`  Subtotal:      ₹${subtotal.toFixed(2)}`);
  if (discount > 0) console.log(`  Discount:     -₹${discount.toFixed(2)}`);
  console.log(`  Tax (GST):     ₹${tax.toFixed(2)}`);
  if (other > 0) console.log(`  Other Charges: ₹${other.toFixed(2)}`);
  console.log(`  ${colors.bold}Total:         ₹${total.toFixed(2)}${colors.reset}`);
  
  // Step 1: Check/Create Customer
  console.log('\n' + '─'.repeat(60));
  log('Step 1: Customer Setup', 'step');
  
  let customerId = invoiceData.customer_id;
  
  if (!customerId) {
    // Try to find existing customer by phone
    try {
      const searchResponse = await fetch(
        `${API_BASE}/customers?search=${encodeURIComponent(invoiceData.primary_phone)}&limit=1`,
        {
          headers: { 'X-Org-Id': ORG_ID }
        }
      );
      
      if (searchResponse.ok) {
        const data = await searchResponse.json();
        if (data.customers && data.customers.length > 0) {
          customerId = data.customers[0].customer_id;
          log(`Found existing customer: ID ${customerId}`, 'success');
        }
      }
    } catch (error) {
      log(`Customer search failed: ${error.message}`, 'warning');
    }
    
    // Create new customer if not found
    if (!customerId) {
      try {
        const customerData = {
          customer_name: invoiceData.customer_name,
          customer_type: invoiceData.customer_type || 'retail',
          primary_phone: invoiceData.primary_phone,
          primary_email: invoiceData.primary_email || '',
          state: invoiceData.state || 'Maharashtra',
          city: invoiceData.city || 'Mumbai',
          credit_limit: invoiceData.credit_limit || 50000,
          credit_period_days: invoiceData.credit_period_days || 30
        };
        
        const createResponse = await fetch(`${API_BASE}/customers`, {
          method: 'POST',
          headers: {
            'X-Org-Id': ORG_ID,
            'Content-Type': 'application/json'
          },
          body: JSON.stringify(customerData)
        });
        
        if (createResponse.ok) {
          const customer = await createResponse.json();
          customerId = customer.customer_id || customer.id || 1;
          log(`Created new customer: ${customerData.customer_name} (ID: ${customerId})`, 'success');
        } else {
          log('Could not create customer, using default ID: 1', 'warning');
          customerId = 1;
        }
      } catch (error) {
        log(`Customer creation failed: ${error.message}`, 'warning');
        customerId = 1;
      }
    }
  }
  
  invoiceData.customer_id = customerId;
  
  // Step 2: Create Invoice
  console.log('\n' + '─'.repeat(60));
  log('Step 2: Invoice Creation', 'step');
  
  try {
    log('Sending invoice data to backend...', 'info');
    
    const response = await fetch(`${API_BASE}/invoices`, {
      method: 'POST',
      headers: {
        'X-Org-Id': ORG_ID,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(invoiceData)
    });
    
    const responseText = await response.text();
    
    if (response.ok) {
      let invoice;
      try {
        invoice = JSON.parse(responseText);
      } catch (e) {
        log('Response parsing issue, but invoice may have been created', 'warning');
        console.log('Raw response:', responseText.substring(0, 200));
        return { status: 'partial_success', response: responseText };
      }
      
      log('Invoice created successfully!', 'success');
      console.log('\n' + colors.bold + 'Invoice Details:' + colors.reset);
      console.log(`  Invoice Number: ${invoice.invoice_number || 'N/A'}`);
      console.log(`  Invoice ID:     ${invoice.invoice_id || 'N/A'}`);
      console.log(`  Customer:       ${invoiceData.customer_name}`);
      console.log(`  Total Amount:   ₹${(invoice.total_amount || total).toFixed(2)}`);
      console.log(`  Status:         ${invoice.invoice_status || 'Created'}`);
      console.log(`  Payment:        ${invoice.payment_status || 'Pending'}`);
      
      // Step 3: Verify Invoice
      if (invoice.invoice_id) {
        console.log('\n' + '─'.repeat(60));
        log('Step 3: Invoice Verification', 'step');
        
        try {
          const verifyResponse = await fetch(
            `${API_BASE}/invoices/${invoice.invoice_id}`,
            {
              headers: { 'X-Org-Id': ORG_ID }
            }
          );
          
          if (verifyResponse.ok) {
            log('Invoice verified in database', 'success');
          } else {
            log('Could not verify invoice (may still be created)', 'warning');
          }
        } catch (error) {
          log(`Verification failed: ${error.message}`, 'warning');
        }
      }
      
      return invoice;
      
    } else {
      log(`Invoice creation failed: HTTP ${response.status}`, 'error');
      console.log('Error details:', responseText.substring(0, 500));
      return null;
    }
    
  } catch (error) {
    log(`Network error: ${error.message}`, 'error');
    return null;
  }
}

// Main execution
async function main() {
  const args = process.argv.slice(2);
  const inputFile = args[0];
  
  console.log('\n' + colors.bold + '🧪 INVOICE CREATION VALIDATION TEST' + colors.reset);
  console.log('═'.repeat(60));
  
  const startTime = Date.now();
  const result = await createInvoiceFromJSON(inputFile);
  const endTime = Date.now();
  
  console.log('\n' + '═'.repeat(60));
  
  if (result) {
    log(`TEST COMPLETED SUCCESSFULLY in ${endTime - startTime}ms`, 'success');
    console.log('\n✨ Invoice has been created and validated!');
  } else {
    log('TEST FAILED - Please check the errors above', 'error');
    process.exit(1);
  }
}

// Run if executed directly
if (require.main === module) {
  main().catch(error => {
    log(`Unexpected error: ${error.message}`, 'error');
    process.exit(1);
  });
}

module.exports = { createInvoiceFromJSON };